### PR TITLE
KBC-2761 migrate Metadata to JSON

### PR DESCRIPTION
**Pull request by martinjunger**
Jira: KBC-2761

Before asking for review make sure that:

## Checklist

- [x] New client method(s) has tests
- [x] Apiary file is updated

## Release

  - [x] I gave the PR a proper label:
    * major (BC break)
    * minor (new feature)
    * patch (backwards compatible fix)
    * no release (just test changes)

### CODE REVIEW

File: src/Keboola/StorageApi/Metadata.php [hunk 1]
```diff
@@ -22,85 +22,81 @@ public function __construct(Client $client)
...
-     * @param $columnId  -- ex: "in.c-bucket.table.column"
+     * @param string $columnId  -- ex: "in.c-bucket.table.column"
      * @return array list of metadata
      */
-    public function listColumnMetadata($columnId)
+    public function listColumnMetadata($columnId): array
     {
-        /** @var array $result */
         $result = $this->client->apiGet("columns/{$columnId}/metadata");
+        assert(is_array($result));
```
Comment (jirkasemmler) [inline src/Keboola/StorageApi/Metadata.php:53]: ty jo nevim jestli bych to delal. to muze dost zmenit chytani vyjimek jinde. Nestaci ten return v anotaci?

File: src/Keboola/StorageApi/Metadata.php [hunk 2]
```diff
@@ -22,85 +22,81 @@ public function __construct(Client $client)
...
-     * @param $metadataId
-     * @return mixed|string
+     * @param string $bucketId
+     * @param string|int $metadataId
      */
-    public function deleteBucketMetadata($bucketId, $metadataId)
+    public function deleteBucketMetadata($bucketId, $metadataId): void
     {
-        return $this->client->apiDelete("buckets/{$bucketId}/metadata/{$metadataId}");
+        $this->client->apiDelete("buckets/{$bucketId}/metadata/{$metadataId}");
```
Comment (jirkasemmler) [inline src/Keboola/StorageApi/Metadata.php:63]: proc z tech DELETE requestu nic nevracime?

**Review by jirkasemmler — CHANGES_REQUESTED** [review]

File: src/Keboola/StorageApi/Metadata.php (hunk 2 above)
Comment (martinjunger) [reply src/Keboola/StorageApi/Metadata.php:63]: 1) protoze to vraci hlavicku `204 No Content` a prazdny string `''`
2) ze spousty DELETE requestu nic nevracime, takze to neni nic divnyho ;-)

File: src/Keboola/StorageApi/Metadata.php (hunk 1 above)
Comment (martinjunger) [reply src/Keboola/StorageApi/Metadata.php:53]: vyhozeny vyjimky  tuhle preskoci, tady se zpracuje jen realny vystup
```php
// \Keboola\StorageApi\Client::request
        try {
            /**
             * @var ResponseInterface $response
             */
            $response = $this->client->request($method, $url, $requestOptions);
        } catch (RequestException $e) {
            $response = $e->getResponse();
            $body = $response ? json_decode((string) $response->getBody(), true) : [];

            if ($response && $response->getStatusCode() == 503) {
                throw new MaintenanceException(isset($body['reason']) ? $body['reason'] : 'Maintenance', $response && $response->hasHeader('Retry-After') ? (string) $response->getHeader('Retry-After')[0] : null, $body);
            }

            throw new ClientException(
                isset($body['error']) ? $body['error'] : $e->getMessage(),
                $response ? $response->getStatusCode() : $e->getCode(),
                $e,
                isset($body['code']) ? $body['code'] : '',
                $body
            );
        }
```

**Comment by martinjunger** [comment]
ℹ️ Rebase

File: src/Keboola/StorageApi/Metadata.php (hunk 1 above)
Comment (zajca) [reply src/Keboola/StorageApi/Metadata.php:53]: imho OK, je to enom aby phpstan nenadával, klidně by tam mohlo byt enom `/** @var array $result */`

**Review by jirkasemmler — APPROVED** [review]
👍

**Comment by martinjunger** [comment]
Vratil jsem JSON na form-data pro Client::postTableMetadata, protoze
> // deprecated use self::postTableMetadataWithColumns() instead - beware of a bit different format of response!
a
> // Keep call with form-data here - call with JSON has different format of response

**Comment by martinjunger** [comment]
[Tag](https://dev.azure.com/keboola-dev/Connection%20build/_build/results?buildId=25869&view=results) prosel 🟢

**Pull request merged**